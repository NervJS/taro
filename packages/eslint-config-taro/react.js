module.exports = Object.assign({}, require('./index'), {
  settings: {
    react: {
      pragma: 'React',
      version: 'detect'
    }
  }
})
