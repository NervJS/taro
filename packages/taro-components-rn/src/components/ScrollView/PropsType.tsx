import * as React from 'react'
import { StyleProp, ViewStyle, ListRenderItem } from 'react-native'

export type ScrollMetrics = {
  contentLength: number;
    dOffset: number;
    dt: number;
    offset: number;
    offsetX: number;
    offsetY: number;
    timestamp: number;
    velocity: number;
    visibleLength: number;
}

export type EventOnScrollToUpper = {
  distanceFromTop: number;
}

export type EventOnScrollToLower = {
  distanceFromEnd: number;
}

export type EventOnScroll = {
  detail: {
    scrollLeft: number;
    scrollTop: number;
    scrollHeight: number;
    scrollWidth: number;
    deltaX: number;
    deltaY: number;
  }
}

export interface ScrollViewState {
  snapScrollTop: number;
  snapScrollLeft: number;
}

export interface ScrollViewProps<ItemT> {
  children?: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  scrollX?: boolean;
  upperThreshold: number;
  lowerThreshold: number;
  scrollTop: number;
  scrollLeft: number;
  scrollWithAnimation?: boolean;
  enableBackToTop: boolean;
  onScrollToUpper?: (evt: EventOnScrollToUpper) => void;
  onScrollToLower?: (evt: EventOnScrollToLower) => void;
  onScroll?: (evt: EventOnScroll) => void;
  // RN 属性
  contentContainerStyle?: StyleProp<ViewStyle>;
  data?: ReadonlyArray<ItemT> | null;
  renderItem?: ListRenderItem<ItemT>;
}
