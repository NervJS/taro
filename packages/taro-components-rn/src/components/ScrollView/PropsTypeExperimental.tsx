import * as React from 'react'
import { StyleProp, ViewStyle, ListRenderItem } from 'react-native'
import { LoadingFooter } from 'react-native-spring-scrollview'

export type ScrollMetrics = {
  contentLength: number;
    dOffset: number;
    dt: number;
    offset: number;
    offsetX: number;
    offsetY: number;
    timestamp: number;
    velocity: number;
    visibleLength: number;
}

export type EventOnScrollToUpper = {
  distanceFromTop: number;
}

export type EventOnScrollToLower = {
  distanceFromEnd: number;
}

export type EventOnScroll = {
  detail: {
    scrollLeft: number;
    scrollTop: number;
    scrollHeight: number;
    scrollWidth: number;
    deltaX: number;
    deltaY: number;
  }
}

export interface ScrollViewState {
  snapScrollTop: number;
  snapScrollLeft: number;
}

export interface ScrollViewProps<ItemT> {
  style?: StyleProp<ViewStyle>;
  data: any[];
  heightForItem: (item:any,index:number) => number;
  renderHeader?: () => React.ReactElement;
  renderFooter?: () => React.ReactElement;
  renderItem: (item:any, index:number) => React.ReactElement

  scrollX?: boolean;
  upperThreshold: number;
  lowerThreshold: number;
  scrollTop: number;
  scrollLeft: number;
  onScrollToUpper?: (evt: EventOnScrollToUpper) => void;
  onScrollToLower?: (evt: EventOnScrollToLower) => void;
  onScroll?: (evt: EventOnScroll) => void;
  // RN 属性
  contentContainerStyle?: StyleProp<ViewStyle>;
}
