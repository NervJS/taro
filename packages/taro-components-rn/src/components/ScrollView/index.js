/**
 * ✔ scrollX(scroll-x): Either-or
 * ✘ scrollY(scroll-y): Either-or
 * ✔ upperThreshold(upper-threshold)
 * ✔ lowerThreshold(lower-threshold)
 * ✔ scrollTop(scroll-top)
 * ✔ scrollLeft(scroll-left)
 * ✘ scroll-into-view
 * ✔ scrollWithAnimation(scroll-with-animation)
 * ✔ enableBackToTop(enable-back-to-top)
 * ✔ onScrollToUpper(bindscrolltoupper)
 * ✔ onScrollToLower(bindscrolltolower)
 * ✔ onScroll(bindscroll)
 *
 * @flow
 */

import * as React from 'react'
import {
  ScrollView,
  StyleSheet,
} from 'react-native'
import { dismemberStyle, omit } from '../../utils'

type Props = {
  children?: React.Node,
  style?: StyleSheet.Styles,
  scrollX?: boolean,
  upperThreshold: number,
  lowerThreshold: number,
  scrollTop: number,
  scrollLeft: number,
  scrollWithAnimation?: boolean,
  enableBackToTop: boolean,
  onScrollToUpper?: Function,
  onScrollToLower?: Function,
  onScroll?: Function,
}

class _ScrollView extends React.Component<Props> {
  // eslint-disable-next-line no-useless-constructor
  constructor (props: Props) {
    super(props)
  }

  props: Props

  static defaultProps = {
    upperThreshold: 50,
    lowerThreshold: 50,
    scrollTop: 0,
    scrollLeft: 0,
    enableBackToTop: false,
  }

  _scrollMetrics = {
    contentLength: 0,
    dOffset: 0,
    dt: 10,
    offset: 0,
    offsetX: 0,
    offsetY: 0,
    timestamp: 0,
    velocity: 0,
    visibleLength: 0,
  }
  _scrollRef: any = null
  _captureScrollRef = (ref: any) => {
    this._scrollRef = ref
  }
  _hasDataChangedSinceEndReached: boolean
  _sentEndForContentLength: number = 0
  _scrollEventThrottle: number = 50
  _hasCallScrollToUpperInRange: boolean = false
  _hasCallScrollToLowerInRange: boolean = false
  _initialScrollIndexTimeout: any

  _selectLength = (metrics: { height: number, width: number }): number => {
    return !this.props.scrollX ? metrics.height : metrics.width
  }

  _selectOffset = (metrics: {x: number, y: number}): number => {
    return !this.props.scrollX ? metrics.y : metrics.x
  }

  _maybeCallOnStartReached = () => {
    const { onScrollToUpper, upperThreshold } = this.props
    const { offset } = this._scrollMetrics
    if (onScrollToUpper && offset < upperThreshold) {
      if (!this._hasCallScrollToUpperInRange) {
        onScrollToUpper({ distanceFromTop: offset })
        this._hasCallScrollToUpperInRange = true
      }
    } else {
      this._hasCallScrollToUpperInRange = false
    }
  }

  _maybeCallOnEndReached = () => {
    const { onScrollToLower, lowerThreshold } = this.props
    const { contentLength, visibleLength, offset } = this._scrollMetrics
    const distanceFromEnd = contentLength - visibleLength - offset
    if (onScrollToLower &&
        distanceFromEnd < lowerThreshold &&
        (this._hasDataChangedSinceEndReached || contentLength !== this._sentEndForContentLength)) {
      if (!this._hasCallScrollToLowerInRange) {
        this._hasDataChangedSinceEndReached = false
        this._hasCallScrollToLowerInRange = true
        this._sentEndForContentLength = this._scrollMetrics.contentLength
        onScrollToLower({ distanceFromEnd })
      }
    } else {
      this._hasCallScrollToLowerInRange = false
    }
  }

  _onContentSizeChange = (width: number, height: number) => {
    this._scrollMetrics.contentLength = this._selectLength({ height, width })
    this._maybeCallOnStartReached()
    this._maybeCallOnEndReached()
  }

  _onScrollEndDrag = (e: Object): void => {
    const { velocity } = e.nativeEvent
    if (velocity) {
      this._scrollMetrics.velocity = this._selectOffset(velocity)
    }
  }

  _onMomentumScrollEnd = (): void => {
    this._scrollMetrics.velocity = 0
  }

  _onLayout = (e: Object) => {
    this._scrollMetrics.visibleLength = this._selectLength(e.nativeEvent.layout)
    this._maybeCallOnStartReached()
    this._maybeCallOnEndReached()
  }

  _onScroll = (e: Object) => {
    const { onScroll } = this.props
    const scrollLeft = e.nativeEvent.contentOffset.x
    const scrollTop = e.nativeEvent.contentOffset.y
    const scrollHeight = e.nativeEvent.contentSize.height
    const scrollWidth = e.nativeEvent.contentSize.width
    onScroll && onScroll({
      detail: {
        scrollLeft,
        scrollTop,
        scrollHeight,
        scrollWidth,
        deltaX: scrollLeft - this._scrollMetrics.offsetX,
        deltaY: scrollTop - this._scrollMetrics.offsetY,
      }
    })

    const timestamp = e.timeStamp
    const visibleLength = this._selectLength(e.nativeEvent.layoutMeasurement)
    const contentLength = this._selectLength(e.nativeEvent.contentSize)
    const offset = this._selectOffset(e.nativeEvent.contentOffset)
    const dt = Math.max(1, timestamp - this._scrollMetrics.timestamp)
    const dOffset = offset - this._scrollMetrics.offset
    const velocity = dOffset / dt
    this._scrollMetrics = {
      contentLength,
      dt,
      dOffset,
      offset,
      offsetX: scrollLeft,
      offsetY: scrollTop,
      timestamp,
      velocity,
      visibleLength,
    }
    this._maybeCallOnStartReached()
    this._maybeCallOnEndReached()
  }

  scrollToOffset = (x: number = 0, y: number = 0) => {
    this._scrollRef.scrollTo({ x, y, animated: !!this.props.scrollWithAnimation })
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps (nextProps: Props) {
    if (nextProps.scrollTop !== this.props.scrollTop || nextProps.scrollLeft !== this.props.scrollLeft) {
      this.scrollToOffset(nextProps.scrollLeft, nextProps.scrollTop)
    }
  }

  // componentWillReceiveProps (nextProps: Props) {
  //   const { data, extraData, getItemCount, maxToRenderPerBatch } = nextProps
  //   if (data !== this.props.data) {
  //     this._hasDataChangedSinceEndReached = true
  //   }
  // }

  componentDidMount () {
    if (this.props.scrollTop || this.props.scrollLeft) {
      this._initialScrollIndexTimeout = setTimeout(() => {
        this.scrollToOffset(this.props.scrollLeft, this.props.scrollTop)
      }, 0)
    }
  }

  componentWillUnmount () {
    this._initialScrollIndexTimeout && clearTimeout(this._initialScrollIndexTimeout)
  }

  render () {
    const {
      children,
      style,
      scrollX,
      enableBackToTop,
    } = this.props

    const dismember = dismemberStyle(style)
    const wrapperStyle = Object.assign(dismember.wrapperStyle, { height: dismember.innerStyle.height })
    const innerStyle = omit(dismember.innerStyle, [ 'height' ])

    return (
      <ScrollView
        horizontal={scrollX}
        onContentSizeChange={this._onContentSizeChange}
        onLayout={this._onLayout}
        onScroll={this._onScroll}
        onScrollEndDrag={this._onScrollEndDrag}
        onMomentumScrollEnd={this._onMomentumScrollEnd}
        ref={this._captureScrollRef}
        scrollEventThrottle={this._scrollEventThrottle}
        scrollsToTop={!!enableBackToTop}
        style={[{ flexGrow: 0 }, wrapperStyle]}
        contentContainerStyle={innerStyle}
      >
        {children}
      </ScrollView>
    )
  }
}

export default _ScrollView
