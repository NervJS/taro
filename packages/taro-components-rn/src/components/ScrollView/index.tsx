/**
 * @see https://facebook.github.io/react-native/docs/scrollview.html
 *
 * 注意事项：
 *   一般地，ScrollView 外面要套一个 View 并在这个 View 上设置高度，否则会出现 ScrollView 撑满外层出现不能滚动的假象
 *   hack: scrollTop 在设置过一次后经过滚动再次设置时失效，因为 state 没变，所以这时可以通过设置一个比 0 小的值
 *
 * ✔ scrollX(scroll-x): Either-or
 * ✘ scrollY(scroll-y): Either-or
 * ✔ upperThreshold(upper-threshold)
 * ✔ lowerThreshold(lower-threshold)
 * ✔ scrollTop(scroll-top)
 * ✔ scrollLeft(scroll-left)
 * ✘ scroll-into-view
 * ✔ scrollWithAnimation(scroll-with-animation)
 * ✔ enableBackToTop(enable-back-to-top)
 * ✔ onScrollToUpper(bindscrolltoupper)
 * ✔ onScrollToLower(bindscrolltolower)
 * ✔ onScroll(bindscroll)
 */

import * as React from 'react'
import {
  ScrollView,
  FlatList,
  NativeSyntheticEvent,
  NativeScrollEvent,
  LayoutChangeEvent,
  StyleSheet,
  ViewStyle
} from 'react-native'
import { omit, noop } from '../../utils'
import { ScrollViewProps, ScrollViewState, ScrollMetrics } from './PropsType'

// const SCROLLVIEW_CONT_STYLE = [
//   // Source code of ScrollView, ['alignItems','justifyContent']
//   'alignItems',
//   'justifyContent',
//   // Other
// ]

class _ScrollView extends React.Component<ScrollViewProps<any>, ScrollViewState> {
  static defaultProps = {
    upperThreshold: 50,
    lowerThreshold: 50,
    scrollTop: 0,
    scrollLeft: 0,
    enableBackToTop: false,
  }

  static getDerivedStateFromProps (props: ScrollViewProps<any>, state: ScrollViewState) {
    return state.snapScrollTop !== props.scrollTop || state.snapScrollLeft !== props.scrollLeft ? {
      snapScrollTop: props.scrollTop,
      snapScrollLeft: props.scrollLeft
    } : null
  }

  state: ScrollViewState = {
    snapScrollTop: 0,
    snapScrollLeft: 0
  }

  _scrollMetrics: ScrollMetrics = {
    contentLength: 0,
    dOffset: 0,
    dt: 10,
    offset: 0,
    offsetX: 0,
    offsetY: 0,
    timestamp: 0,
    velocity: 0,
    visibleLength: 0,
  }
  $scrollView = React.createRef<any>()
  _hasDataChangedSinceEndReached: boolean
  _sentEndForContentLength: number = 0
  _scrollEventThrottle: number = 50
  _hasCallScrollToUpperInRange: boolean = false
  _hasCallScrollToLowerInRange: boolean = false
  _initialScrollIndexTimeout: any

  _selectLength = (metrics: { height: number, width: number }): number => {
    return !this.props.scrollX ? metrics.height : metrics.width
  }

  _selectOffset = (metrics: {x: number, y: number}): number => {
    return !this.props.scrollX ? metrics.y : metrics.x
  }

  _maybeCallOnStartReached = (): void => {
    const { onScrollToUpper, upperThreshold } = this.props
    const { offset } = this._scrollMetrics
    if (onScrollToUpper && offset < upperThreshold) {
      if (!this._hasCallScrollToUpperInRange) {
        onScrollToUpper({ distanceFromTop: offset })
        this._hasCallScrollToUpperInRange = true
      }
    } else {
      this._hasCallScrollToUpperInRange = false
    }
  }

  _maybeCallOnEndReached = (): void => {
    const { onScrollToLower, lowerThreshold } = this.props
    const { contentLength, visibleLength, offset } = this._scrollMetrics
    const distanceFromEnd = contentLength - visibleLength - offset
    if (onScrollToLower &&
        distanceFromEnd < lowerThreshold &&
        (this._hasDataChangedSinceEndReached || contentLength !== this._sentEndForContentLength)) {
      if (!this._hasCallScrollToLowerInRange) {
        this._hasDataChangedSinceEndReached = false
        this._hasCallScrollToLowerInRange = true
        this._sentEndForContentLength = this._scrollMetrics.contentLength
        onScrollToLower({ distanceFromEnd })
      }
    } else {
      this._hasCallScrollToLowerInRange = false
    }
  }

  _onContentSizeChange = (width: number, height: number): void => {
    this._scrollMetrics.contentLength = this._selectLength({ height, width })
    // this._maybeCallOnStartReached()
    // this._maybeCallOnEndReached()
  }

  _onScrollEndDrag = (e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { velocity } = e.nativeEvent
    if (velocity) {
      this._scrollMetrics.velocity = this._selectOffset(velocity)
    }
  }

  _onMomentumScrollEnd = (): void => {
    this._scrollMetrics.velocity = 0
  }

  _onLayout = (e: LayoutChangeEvent): void => {
    this._scrollMetrics.visibleLength = this._selectLength(e.nativeEvent.layout)
    // this._maybeCallOnStartReached()
    // this._maybeCallOnEndReached()
  }

  _onScroll = (e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { onScroll = noop } = this.props
    const scrollLeft: number = e.nativeEvent.contentOffset.x
    const scrollTop: number = e.nativeEvent.contentOffset.y
    const scrollHeight: number = e.nativeEvent.contentSize.height
    const scrollWidth: number = e.nativeEvent.contentSize.width
    onScroll({
      detail: {
        scrollLeft,
        scrollTop,
        scrollHeight,
        scrollWidth,
        deltaX: scrollLeft - this._scrollMetrics.offsetX,
        deltaY: scrollTop - this._scrollMetrics.offsetY,
      }
    })

    const timestamp: number = e.timeStamp
    const visibleLength: number = this._selectLength(e.nativeEvent.layoutMeasurement)
    const contentLength: number = this._selectLength(e.nativeEvent.contentSize)
    const offset: number = this._selectOffset(e.nativeEvent.contentOffset)
    const dt: number = Math.max(1, timestamp - this._scrollMetrics.timestamp)
    const dOffset: number = offset - this._scrollMetrics.offset
    const velocity: number = dOffset / dt
    this._scrollMetrics = {
      contentLength,
      dt,
      dOffset,
      offset,
      offsetX: scrollLeft,
      offsetY: scrollTop,
      timestamp,
      velocity,
      visibleLength,
    }
    this._maybeCallOnStartReached()
    this._maybeCallOnEndReached()
  }

  scrollToOffset = (x: number = 0, y: number = 0): void => {
    const { scrollX, data, renderItem } = this.props
    const node = this.$scrollView.current
    if (node) {
      if (data && renderItem) {
        (node as FlatList<any>).scrollToOffset({ offset: scrollX ? x : y, animated: !!this.props.scrollWithAnimation })
      } else {
        (node as ScrollView).scrollTo({ x, y, animated: !!this.props.scrollWithAnimation })
      }
    }
  }

  componentDidMount () {
    if (this.state.snapScrollTop || this.state.snapScrollLeft) {
      this._initialScrollIndexTimeout = setTimeout(() => {
        this.scrollToOffset(this.state.snapScrollLeft, this.state.snapScrollTop)
      }, 0)
    }
  }

  getSnapshotBeforeUpdate (prevProps: ScrollViewProps<any>, prevState: ScrollViewState) {
    return prevState.snapScrollTop !== this.state.snapScrollTop || prevState.snapScrollLeft !== this.state.snapScrollLeft
  }

  componentDidUpdate (prevProps: ScrollViewProps<any>, prevState: ScrollViewState, snapshot: boolean) {
    if (snapshot) {
      this.scrollToOffset(this.state.snapScrollLeft, this.state.snapScrollTop)
    }
  }

  componentWillUnmount () {
    this._initialScrollIndexTimeout && clearTimeout(this._initialScrollIndexTimeout)
  }

  render () {
    const {
      children,
      style,
      scrollX,
      enableBackToTop,
      contentContainerStyle,
      data,
      renderItem,
    } = this.props

    const flattenStyle: ViewStyle & { [key: string]: any } = StyleSheet.flatten(style)
    const wrapperStyle: ViewStyle = omit(flattenStyle, [
      'alignItems',
      'justifyContent'
    ])
    const _contentContainerStyle: ViewStyle & { [key: string]: any } = {}
    if (flattenStyle) {
      flattenStyle.alignItems && (_contentContainerStyle.alignItems = flattenStyle.alignItems)
      flattenStyle.justifyContent && (_contentContainerStyle.justifyContent = flattenStyle.justifyContent)
    }

    const scrollElementProps = {
      horizontal: scrollX,
      onContentSizeChange: this._onContentSizeChange,
      onLayout: this._onLayout,
      onScroll: this._onScroll,
      onScrollEndDrag: this._onScrollEndDrag,
      onMomentumScrollEnd: this._onMomentumScrollEnd,
      scrollEventThrottle: this._scrollEventThrottle,
      scrollsToTop: !!enableBackToTop,
      style: wrapperStyle,
      contentContainerStyle: [_contentContainerStyle, contentContainerStyle],
      ...omit(this.props, [
        // props
        'style',
        'scrollX',
        'upperThreshold',
        'lowerThreshold',
        'scrollTop',
        'scrollLeft',
        'scrollWithAnimation',
        'enableBackToTop',
        'onScrollToUpper',
        'onScrollToLower',
        'onScroll',
        'contentContainerStyle',
        // SProps
        'horizontal',
        'onContentSizeChange',
        'onLayout',
        'onScroll',
        'onScrollEndDrag',
        'onMomentumScrollEnd',
        'scrollEventThrottle',
        'scrollsToTop',
        'style',
        'contentContainerStyle',
        'data',
        'renderItem',
        'keyExtractor',
      ]),
      ref: this.$scrollView
    }

    return data && renderItem ? (
      <FlatList
        {...scrollElementProps}
        data={data}
        renderItem={renderItem}
        keyExtractor={(item, index) => index + ''}
      />
    ) : (
      <ScrollView
        {...scrollElementProps}
      >
        {children}
      </ScrollView>
    )
  }
}

export default _ScrollView
