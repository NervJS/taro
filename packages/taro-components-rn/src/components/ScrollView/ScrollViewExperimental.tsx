/**
 * ScrollViewExperimental
 */

import * as React from 'react'
import {
  ScrollView,
  FlatList,
  NativeSyntheticEvent,
  NativeScrollEvent,
  LayoutChangeEvent,
  StyleSheet,
  ViewStyle,
  View,
  Text
} from 'react-native'
import { WaterfallList } from 'react-native-largelist-v3'
import { omit, noop } from '../../utils'
import { ScrollViewProps, ScrollViewState, ScrollMetrics } from './PropsTypeExperimental'

class _ScrollView extends React.Component<ScrollViewProps<any>, ScrollViewState> {
  static defaultProps = {
    upperThreshold: 50,
    lowerThreshold: 50,
    scrollTop: 0,
    scrollLeft: 0,
  }

  static getDerivedStateFromProps (props: ScrollViewProps<any>, state: ScrollViewState) {
    return state.snapScrollTop !== props.scrollTop || state.snapScrollLeft !== props.scrollLeft ? {
      snapScrollTop: props.scrollTop,
      snapScrollLeft: props.scrollLeft
    } : null
  }

  state: ScrollViewState = {
    snapScrollTop: 0,
    snapScrollLeft: 0
  }

  _scrollMetrics: ScrollMetrics = {
    contentLength: 0,
    dOffset: 0,
    dt: 10,
    offset: 0,
    offsetX: 0,
    offsetY: 0,
    timestamp: 0,
    velocity: 0,
    visibleLength: 0,
  }
  $scrollView = React.createRef<WaterfallList<any>>()
  _hasDataChangedSinceEndReached: boolean
  _sentEndForContentLength: number = 0
  _scrollEventThrottle: number = 50
  _hasCallScrollToUpperInRange: boolean = false
  _hasCallScrollToLowerInRange: boolean = false
  _initialScrollIndexTimeout: any

  _selectLength = (metrics: { height: number, width: number }): number => {
    return !this.props.scrollX ? metrics.height : metrics.width
  }

  _selectOffset = (metrics: {x: number, y: number}): number => {
    return !this.props.scrollX ? metrics.y : metrics.x
  }

  _onContentSizeChange = (width: number, height: number): void => {
    this._scrollMetrics.contentLength = this._selectLength({ height, width })
  }

  _onScrollEndDrag = (e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { velocity } = e.nativeEvent
    if (velocity) {
      this._scrollMetrics.velocity = this._selectOffset(velocity)
    }
  }

  _onMomentumScrollEnd = (): void => {
    this._scrollMetrics.velocity = 0
  }

  _onLayout = (e: LayoutChangeEvent): void => {
    this._scrollMetrics.visibleLength = this._selectLength(e.nativeEvent.layout)
  }

  _maybeCallOnStartReached = (): void => {
    const { onScrollToUpper, upperThreshold } = this.props
    const { offset } = this._scrollMetrics
    if (onScrollToUpper && offset < upperThreshold) {
      if (!this._hasCallScrollToUpperInRange) {
        onScrollToUpper({ distanceFromTop: offset })
        this._hasCallScrollToUpperInRange = true
      }
    } else {
      this._hasCallScrollToUpperInRange = false
    }
  }

  _maybeCallOnEndReached = (): void => {
    const { onScrollToLower, lowerThreshold } = this.props
    const { contentLength, visibleLength, offset } = this._scrollMetrics
    const distanceFromEnd = contentLength - visibleLength - offset
    if (onScrollToLower &&
        distanceFromEnd < lowerThreshold &&
        (this._hasDataChangedSinceEndReached || contentLength !== this._sentEndForContentLength)) {
      if (!this._hasCallScrollToLowerInRange) {
        this._hasDataChangedSinceEndReached = false
        this._hasCallScrollToLowerInRange = true
        this._sentEndForContentLength = this._scrollMetrics.contentLength
        onScrollToLower({ distanceFromEnd })
      }
    } else {
      this._hasCallScrollToLowerInRange = false
    }
  }

  _onScroll = (e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { onScroll = noop } = this.props
    const scrollLeft: number = e.nativeEvent.contentOffset.x
    const scrollTop: number = e.nativeEvent.contentOffset.y
    const scrollHeight: number = e.nativeEvent.contentSize.height
    const scrollWidth: number = e.nativeEvent.contentSize.width
    onScroll({
      detail: {
        scrollLeft,
        scrollTop,
        scrollHeight,
        scrollWidth,
        deltaX: scrollLeft - this._scrollMetrics.offsetX,
        deltaY: scrollTop - this._scrollMetrics.offsetY,
      }
    })

    const timestamp: number = e.timeStamp
    const visibleLength: number = this._selectLength(e.nativeEvent.layoutMeasurement)
    const contentLength: number = this._selectLength(e.nativeEvent.contentSize)
    const offset: number = this._selectOffset(e.nativeEvent.contentOffset)
    const dt: number = Math.max(1, timestamp - this._scrollMetrics.timestamp)
    const dOffset: number = offset - this._scrollMetrics.offset
    const velocity: number = dOffset / dt
    this._scrollMetrics = {
      contentLength,
      dt,
      dOffset,
      offset,
      offsetX: scrollLeft,
      offsetY: scrollTop,
      timestamp,
      velocity,
      visibleLength,
    }
    this._maybeCallOnStartReached()
    this._maybeCallOnEndReached()
  }

  scrollToOffset = (x: number = 0, y: number = 0): void => {
    const node = this.$scrollView.current
    node && node.scrollTo({ x, y })
  }

  componentDidMount () {
    if (this.state.snapScrollTop || this.state.snapScrollLeft) {
      this._initialScrollIndexTimeout = setTimeout(() => {
        this.scrollToOffset(this.state.snapScrollLeft, this.state.snapScrollTop)
      }, 0)
    }
  }

  getSnapshotBeforeUpdate (prevProps: ScrollViewProps<any>, prevState: ScrollViewState) {
    return prevState.snapScrollTop !== this.state.snapScrollTop || prevState.snapScrollLeft !== this.state.snapScrollLeft
  }

  componentDidUpdate (prevProps: ScrollViewProps<any>, prevState: ScrollViewState, snapshot: boolean) {
    if (snapshot) {
      this.scrollToOffset(this.state.snapScrollLeft, this.state.snapScrollTop)
    }
  }

  componentWillUnmount () {
    this._initialScrollIndexTimeout && clearTimeout(this._initialScrollIndexTimeout)
  }

  render () {
    const {
      style,
      data,
      heightForItem,
      renderHeader,
      renderFooter,
      renderItem,
      contentContainerStyle,
    } = this.props

    const flattenStyle: ViewStyle & { [key: string]: any } = StyleSheet.flatten(style)
    const wrapperStyle: ViewStyle = omit(flattenStyle, [
      'alignItems',
      'justifyContent'
    ])
    const _contentContainerStyle: ViewStyle & { [key: string]: any } = {}
    if (flattenStyle) {
      flattenStyle.alignItems && (_contentContainerStyle.alignItems = flattenStyle.alignItems)
      flattenStyle.justifyContent && (_contentContainerStyle.justifyContent = flattenStyle.justifyContent)
    }

    return (
      <WaterfallList
        renderHeader={renderHeader}
        renderFooter={renderFooter}
        heightForItem={heightForItem}
        numColumns={2}
        renderItem={renderItem}
        // 参考 https://github.com/bolan9999/react-native-spring-scrollview/blob/master/src/NormalFooter.js
        // loadingFooter={loadingFooter}
        data={data}
        onScroll={this._onScroll}
        style={wrapperStyle}
        contentStyle={_contentContainerStyle}
        ref={this.$scrollView}
      />
    )
  }
}

export default _ScrollView
