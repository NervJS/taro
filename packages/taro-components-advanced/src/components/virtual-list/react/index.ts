import { ScrollView, View } from '@tarojs/components'
import React from 'react'

import List from './list'
import outerWrapper from './wrapper'

import type { VirtualListProps } from '../'
import type { IProps } from '../preset'

const VirtualList = React.forwardRef(function VirtualList (props: VirtualListProps, ref) {
  const {
    direction = 'ltr',
    outerElementType = ScrollView,
    innerElementType = View,
    itemElementType = View,
    initialScrollOffset = 0,
    overscanCount = 1,
    ...rest
  } = props as IProps

  if ('children' in rest) {
    console.warn('Taro(VirtualList): children props have been deprecated. ' + 'Please use the item props instead.')
    rest.item = rest.children as IProps['item']
  }
  if (rest.item instanceof Array) {
    console.warn('Taro(VirtualList): item should not be an array')
    rest.item = rest.item[0]
  }
  return React.createElement(List, {
    ref,
    ...rest,
    outerElementType,
    itemElementType,
    innerElementType,
    direction,
    initialScrollOffset,
    overscanCount,
    outerWrapper,
  })
})

export default VirtualList
