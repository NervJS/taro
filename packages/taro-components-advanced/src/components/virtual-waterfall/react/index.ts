import { ScrollView, View } from '@tarojs/components'
import React from 'react'

import Waterfall from './waterfall'
import outerWrapper from './wrapper'

import type { VirtualWaterfallProps } from '../'
import type { IProps } from '../preset'

const VirtualWaterfall = React.forwardRef(function VirtualWaterfall (props: VirtualWaterfallProps, ref) {
  const {
    outerElementType = ScrollView,
    innerElementType = View,
    itemElementType = View,
    initialScrollOffset = 0,
    overscanDistance = 50,
    ...rest
  } = props as IProps

  if ('children' in rest) {
    console.warn('Taro(VirtualWaterfall): children props have been deprecated. ' + 'Please use the item props instead.')
    rest.item = rest.children as IProps['item']
  }
  if (rest.item instanceof Array) {
    console.warn('Taro(VirtualWaterfall): item should not be an array')
    rest.item = rest.item[0]
  }
  return React.createElement(Waterfall, {
    ref,
    ...rest,
    outerElementType,
    itemElementType,
    innerElementType,
    initialScrollOffset,
    overscanDistance,
    outerWrapper,
  })
})

export default VirtualWaterfall
