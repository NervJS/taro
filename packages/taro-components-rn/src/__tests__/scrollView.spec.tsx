import * as React from 'react'
import { render } from '@testing-library/react-native'
import ScrollView from '../components/ScrollView'

describe('ScrollView', () => {
  it('ScrollView render', () => {
    const wrapper = render(<ScrollView>scroll me</ScrollView>).toJSON()
    expect(wrapper).toMatchSnapshot()
  })
})
