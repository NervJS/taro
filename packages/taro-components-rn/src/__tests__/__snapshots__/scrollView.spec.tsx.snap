// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollView ScrollView render 1`] = `
<RCTScrollView
  contentContainerStyle={
    Array [
      Object {},
      undefined,
    ]
  }
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollEndDrag={[Function]}
  scrollEventThrottle={50}
  scrollsToTop={false}
  style={Object {}}
>
  <View>
    scroll me
  </View>
</RCTScrollView>
`;
